vllm/entrypoints/chat_utils.py

```diff
@@ -112,44 +193,39 @@ def load_chat_template(
     return resolved_chat_template
 
 
-@lru_cache(maxsize=None)
-def _mm_token_str(model_config: ModelConfig, tokenizer: AnyTokenizer,
-                  modality: Literal["image", "audio"]) -> Optional[str]:
-    # TODO: Let user specify how to insert image tokens into prompt
-    # (similar to chat template)
-    model_type = model_config.hf_config.model_type
-    if modality == "image":
-        if model_type == "phi3_v":
-            # Workaround since this token is not defined in the tokenizer
-            return "<|image_1|>"
-        if model_type == "minicpmv":
-            return "(<image>./</image>)"
-        if model_type in ("blip-2", "chatglm", "fuyu", "paligemma"):
-            # These models do not use image tokens in the prompt
-            return None
-        if model_type.startswith("llava"):
-            return tokenizer.decode(model_config.hf_config.image_token_index)
-        if model_type in ("chameleon", "internvl_chat"):
-            return "<image>"
-
-        raise TypeError(f"Unknown model type: {model_type}")
-    elif modality == "audio":
-        if model_type == "ultravox":
-            return "<|reserved_special_token_0|>"
-        raise TypeError(f"Unknown model type: {model_type}")
-    else:
-        raise TypeError(f"Unknown modality: {modality}")
-
-
 # TODO: Let user specify how to insert multimodal tokens into prompt
 # (similar to chat template)
-def _get_full_multimodal_text_prompt(placeholder_token_str: str,
+def _get_full_multimodal_text_prompt(placeholders: List[str],
                                      text_prompt: str) -> str:
     """Combine multimodal prompts for a multimodal language model"""
 
+    # Look through the text prompt so that we don't add placeholders for
+    # items that already have them.
+    prompt_fragments = [text_prompt]
+    missing_placeholders: List[str] = []
+    for placeholder in placeholders:
+        for fragment_index, fragment in enumerate(prompt_fragments):
+            index = fragment.find(placeholder)
+            if index >= 0:
+                # This part of the text prompt already has a placeholder.
+                # Remove it from the text prompt fragments so that we don't
+                # consider it for later placeholders.
+                prompt_fragments.pop(fragment_index)
+                before = fragment[:index]
+                after = fragment[index + len(placeholder):]
+                if before:
+                    prompt_fragments.insert(fragment_index, before)
+                if after:
+                    prompt_fragments.insert(fragment_index + 1, after)
+                break
```

Comment:
I'm a little confused about what we're doing here. Isn't `text_prompt` already a string (and thus `prompt_fragments` is a list of one string)?

Did you mean to iterate over `texts` instead of `text_prompt = "\n".join(texts)`?